### PR TITLE
petsc, p4est: strict dependencies

### PR DESCRIPTION
**Pull request by cburstedde**
###### Motivation for this change

I'm now using `strictDeps = true;` to enforce cleanliness in petsc, p4est, p4est-sc. Works fine, but:

 * MPI seems to be a native dep. This is fine for `mpicc`, but I'd think `mpirun` would be a runtime dep. Would it make sense to split openmpi and mpich into two packages to separate compiler and runtime?
 * OpenMPI requires OpenSSH, apparently as a native dep. Wondering whether it should not be a runtime dep, but I don't know how deeply it is compiled in.

These are issues outside of this PR, and everything works since tests are only run when host == build, but I thought I'd ask.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cburstedde** [comment]
nixpkgs-review pr 130156:

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/130156

6 packages built:
getdp p4est p4est-dbg p4est-sc p4est-sc-dbg petsc

**Comment by SuperSandro2000** [comment]
> Would it make sense to split openmpi and mpich into two packages to separate compiler and runtime?

we could output split it.

File: pkgs/development/libraries/science/math/p4est-sc/default.nix [hunk 1]
```diff
@@ -22,11 +22,12 @@ stdenv.mkDerivation {
     sha256 = "14vm0b162jh8399pgpsikbwq4z5lkrw9vfzy3drqykw09n6nc53z";
   };
 
+  strictDeps = true;
   nativeBuildInputs = [ autoreconfHook pkg-config ];
-  propagatedBuildInputs = [ zlib ]
-    ++ lib.optional mpiSupport mpi
-    ++ lib.optional isOpenmpi openssh
+  propagatedNativeBuildInputs =
+    lib.optional mpiSupport mpi ++ lib.optional isOpenmpi openssh
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/p4est-sc/default.nix:28]:
```suggestion
  propagatedNativeBuildInputs = lib.optional mpiSupport mpi
    ++ lib.optional isOpenmpi openssh
```

**Comment by SuperSandro2000** [comment]
> OpenMPI requires OpenSSH, apparently as a native dep. Wondering whether it should not be a runtime dep, but I don't know how deeply it is compiled in.

I think we should wrap it instead to avoid adding it to multiple packages.

**Pull request merged**